Add images user journey (#2991)
viamrobotics/docs: docs/use-cases/deploy-ml.md

## Train and test a machine learning (ML) model

{{< table >}}
{{< tablestep >}}
{{<imgproc src="/services/ml/train.svg" class="fill alignleft" style="max-width: 150px" declaredimensions=true alt="Train models">}}
**1. Train an ML model [_(i)_](/services/ml/train-model/)**

In the Viam app, navigate to your list of [**DATASETS**](https://app.viam.com/services/data/datasets) and select the one you want to train on.
Click **Train model** and follow the prompts.

{{< /tablestep >}}
{{< tablestep >}}
{{<imgproc src="/registry/upload-module.svg" class="fill alignleft" style="max-width: 150px" declaredimensions=true alt="Train models">}}
**2. Deploy your ML model**

Configure the built-in [ML model service](/services/ml/deploy/) to deploy and run the model.
Once you've added the ML model service to your machine, choose your newly-trained model from the dropdown menu in the ML model service's configuration card.

{{< /tablestep >}}
{{< tablestep >}}
{{<imgproc src="/services/icons/vision.svg" class="fill alignleft" style="max-width: 150px" declaredimensions=true alt="Configure a service">}}
**3. Configure an <code>mlmodel</code> vision service [_(i)_](/services/vision/)**

> You can think of the vision service as the bridge between the ML model service and the output from your camera.

Reviewer: I don't think a bridge adds enough meaning as a metaphor here - so adding a bit extra
```suggestion
You can think of the vision service as the bridge between the ML model service and the output from your camera.
The vision service applies the ML model to the images stream from the camera.
```

Author: I can just get rid of it
```suggestion
The vision service takes the the ML model and applies it to the stream of images from your camera.
```

---
title: "How to train and deploy ML/computer vision models"
linkTitle: "Train computer vision models"
weight: 20
type: "docs"
tags: ["vision", "data", "services"]
images: ["/services/ml/collect.svg"]
description: "Collect images and do interesting things with computer vision, ML, and webhooks."
---

You can use Viam's built-in tools to train a machine learning (ML) model on your images and then deploy computer vision on your machines.

![Diagram of the camera component to data management service to ML model service to vision service pipeline.](/use-cases/ml-vision-diagram.png)

For example, you can train a model to recognize your dog and detect whether they are sitting or standing.
Then, you can configure your machine to [capture images](/use-cases/image-data/) only when your dog is in the camera frame so you don't capture hundreds of photos of an empty room.
You can then get even more image data of your dog and improve your ML model by training it on the larger dataset.

You can do all of this using the [Viam app](https://app.viam.com) user interface.
You will not need to write any code.

> <div style="border:1px solid; box-shadow:5px 5px 0 0;padding:1rem; margin:1rem; background-color:#f5f5f5">
> 
> **In this page:**
> 
> 1. [Create a dataset and label data](#create-a-dataset-and-label-data)
> 2. [Train and test a machine learning (ML) model](#train-and-test-a-machine-learning-ml-model)
> 
> </div>

Reviewer: ```suggestion
{{< alert title="In this page" color="tip" >}}

1. [Create a dataset and label data](#create-a-dataset-and-label-data)
2. [Train and test a machine learning (ML) model](#train-and-test-a-machine-learning-ml-model)

{{< /alert >}}
```

Author: If we are okay overloading the alert color code versus creating a new one, sure! Will apply. Perhaps we should consider adding a purple color or something for learning goals and contents.

## Create a dataset and label data

{{< table >}}
{{< tablestep >}}
{{<imgproc src="/services/icons/data-capture.svg" class="fill alignleft" style="max-width: 150px" declaredimensions=true alt="Collect data">}}
**1. Collect images**

Start by collecting images from your cameras and syncing it to the [Viam app](https://app.viam.com).
See [Collect image data and sync it to the cloud](/use-cases/image-data/#collect-image-data-and-sync-it-to-the-cloud) for instructions.

> {{% alert title="Tip" color="tip" %}}

Reviewer: Please move this under the steps for creating and labeling data. You might need to adjust the text a little but I think since it's more of an aside anyway it can go at the bottom of this segment

Author: Even though they'd need to do this during this step, before collecting the data?

> <br>
> {{% alert title="Tip" color="tip" %}}
> To keep your data organized, configure a tag in your data management service config panel.
> This tag will be applied to all data synced from that machine.
> If you apply the same tag to all data gathered from all machines that you want to use in your dataset, you can filter by that tag in the Viam app **DATA** tab to make the next steps easier.
> 
> This is not required, since you can use other filters like time or machine ID in the **DATA** tab to isolate your data.
> {{% /alert %}}
> 

Author: ```suggestion
```

{{< /tablestep >}}
{{< tablestep >}}
{{<imgproc src="/services/ml/collect.svg" class="fill alignleft" style="max-width: 150px" declaredimensions=true alt="Label data">}}
**2. Label your images [_(i)_](/services/data/dataset/)**

Once you have enough images of the objects you'd like to classify, use the interface on the **DATA** tab to label your data.
If you want to train an image classifier, use image tags.
For an object detector, use bounding boxes.

{{< /tablestep >}}
{{< tablestep >}}
{{<imgproc src="/services/ml/label.svg" class="fill alignleft" style="max-width: 150px" declaredimensions=true alt="Label data">}}
**2. Create a dataset [_(i)_](/services/data/dataset/)**

Use the interface on the **DATA** tab (or the [`viam data dataset add` command](/cli/#data)) to add all images you want to train the model on to a dataset.

{{< /tablestep >}}

> {{< /table >}}

Author: ```suggestion
{{< /table >}}

{{% alert title="Tip" color="tip" %}}
To keep your data organized, you can configure a tag in your data management service config panel.
This tag will be applied to all data synced from that machine in the future.
If you apply the same tag to all data gathered from all machines that you want to use in your dataset, you can filter by that tag in the Viam app **DATA** tab, or when querying data.

This is not required, since you can use other filters like time or machine ID in the **DATA** tab to isolate your data.
{{% /alert %}}
```
